feat(category_theory): adjoint equivalences and limits under equivalences

fpvandoorn opened the pull request:
Main features:
* Define equivalences to be adjoint equivalences.
  * Show that one triangle law implies the other for equivalences
  * Construct an adjoint equivalence from a (non-adjoint) equivalence
* Prove that having a limit of shape `J` is preserved under an equivalence `J ≌ J'`.
* Show that the opposite of a discrete category is discrete.
* Put `nat_iso.app` in the `iso` namespace, so that we can write `α.app` for `α : F ≅ G`.
* Add notation `𝟭` for `functor.id` (no backslash-input for VScode yet)
* Add some basic lemmas about equivalences and isomorphisms.
* Move some lemmas from `nat_trans` to `functor_category` and state them using `F ⟶ G` instead of `nat_trans F G` (maybe these files should just be merged?)
* Some small fixes

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

fpvandoorn committed:
feat(category_theory): adjoint equivalences and limits

* Define equivalences to be adjoint equivalences.
  * Show that one triangle law implies the other for equivalences
  * Prove that having a limit of shape `J` is preserved under an equivalence `J ≌ J'`.
  * Construct an adjoint equivalence from a (non-adjoint) equivalence
* Put `nat_iso.app` in the `iso` namespace, so that we can write `α.app` for `α : F ≅ G`.
* Add some basic lemmas about equivalences, isomorphisms.
* Move some lemmas from `nat_trans` to `functor_category` and state them using `F ⟶ G` instead of `nat_trans F G` (maybe these files should just be merged?)
* Some small tweaks, improvements

fpvandoorn committed:
opposite of discrete is discrete

This also shows that C^op has coproducts if C has products and vice versa
Also fix rebase errors

fpvandoorn committed:
fix error (I don't know what caused this to break)

jcommelin reviewed (commented):
Two thoughts:
* This PR is making quite a lot of changes at the same time. It would be easier to review if it was split in three or four smaller refactoring PRs
* During my first quick glance, I didn't see a definition that turns an adjoint equivalence into an adjunction. Could you please add this?

jcommelin committed:
Use tidy a bit more

fpvandoorn committed:
construct an adjunction from an equivalence

add notation `⊣` for an adjunction.
make some arguments of adjunction constructors implicit

jcommelin commented on src/category_theory/adjunction.lean:
```diff
@@ -29,6 +29,8 @@ structure adjunction (F : C ⥤ D) (G : D ⥤ C) :=
 (hom_equiv_unit' : Π {X Y f}, (hom_equiv X Y) f = (unit : _ ⟶ _).app X ≫ G.map f . obviously)
 (hom_equiv_counit' : Π {X Y g}, (hom_equiv X Y).symm g = F.map g ≫ counit.app Y . obviously)
 
+infix ` ⊣ `:15 := adjunction
```
Good idea to introduce this notation. Could you also change the other examples of adjunctions that are out there, so that they also use this notations? Thanks!

fpvandoorn replied:
ok!

fpvandoorn replied:
Shall I also use `notation `𝟭` := functor.id` consistently throughout the category theory library?

jcommelin replied:
Hmm... I guess that's more work (even if it is trivial) and clutters up the PR. I'd rather have that in a separate PR.

jcommelin replied:
Also... maybe we should first think a bit about making that global notation...

fpvandoorn committed:
use adjunction notation

kim-em commented on src/category_theory/discrete_category.lean:
```diff
@@ -14,9 +14,16 @@ def discrete (α : Type u₁) := α
 instance discrete_category (α : Type u₁) : small_category (discrete α) :=
 { hom  := λ X Y, ulift (plift (X = Y)),
-  id   := by tidy,
+  id   := λ X, ulift.up (plift.up rfl),
```
Why did you replace `tidy` here? I don't really mind, but it seems we should either use `tidy` for both `id` and `comp`, or for neither.

fpvandoorn replied:
When `tidy` sees a morphism `f : X ⟶ Y` in a discrete category, it will do cases on this multiple times, until it finally reduces it to a bunch of constructors to get a morphism of type `X ⟶ X`. I added a rewrite lemma (`id_def`) that rewrites this to the identity morphism `𝟙 X`. It is especially nice if this rewrite rule is by `rfl`, because then `dsimp` can rewrite with it. The term `tidy` gave of the identity morphism was not definitionally equal.

I could give `comp` explicitly, but it's not necessary: after rewriting all morphisms to `𝟙 X`, tidy can rewrite `𝟙 X ≫ 𝟙 X` to `𝟙 X`.

fpvandoorn replied:
I added an explicit definition of `comp`.

kim-em committed:
formatting

kim-em committed:
do adjointify_η as a natural iso directly, to avoid checking naturality

kim-em committed:
tersifying

fpvandoorn committed:
fix errors, a bit of cleanup

fpvandoorn committed:
fix elements.lean

rwbarton commented on src/category_theory/equivalence.lean:
```diff
@@ -1,156 +1,256 @@
+
+section
+open category_theory
+variables {F : C ⥤ D} {G : D ⥤ C} (η : 𝟭 C ≅ F ⋙ G) (ε : G ⋙ F ≅ 𝟭 D)
```
As Lean forces the auxiliary definitions to go first, it might help to add a one-line explanatory comment somewhere around here, like `-- Convert an arbitrary equivalence to a half-adjoint equivalence.`

rwbarton commented on src/category_theory/equivalence.lean:
```diff
@@ -1,156 +1,256 @@
+    intros X Y g, dsimp at *, simp at *, dsimp at *,
+    slice_rhs 2 3 { erw [is_iso.hom_inv_id] }, rw [id_comp],
+    slice_rhs 1 2 { erw [is_iso.hom_inv_id] }, rw [id_comp, assoc] },
+  fapply of_components, exact feef_iso_id e f,
```
I would tend to use `{ ... }` also for the last goal when it is parallel to an earlier goal, and try to make the two proofs look similar.

fpvandoorn committed:
Merge remote-tracking branch 'blessed/master' into limits-isos

fpvandoorn committed:
fix error, address comments

Commit:
Merge branch 'master' into 'limits-isos'
